### PR TITLE
Fix compiler warnings when compiling with nvc++

### PR DESCRIPTION
**Pull request by masterleinad**
Split from #4196. In particular, the `half_t` operations must use references.

### CODE REVIEW

**Comment by masterleinad** [comment]
@e10harvey Could you contribute a test for the (presumably) faulty `half_t` operations?

File: core/unit_test/TestPolicyConstruction.hpp [hunk 1]
```diff
@@ -795,7 +795,7 @@ TEST(TEST_CATEGORY, desired_occupancy_empty_base_optimization) {
   static_assert(sizeof(decltype(policy)) == 1, "");
   static_assert_dummy_policy_must_be_size_one<sizeof(decltype(policy))>
       _assert1{};
-  (void)_assert1;  // avoid unused variable warning
+  (void)&_assert1;  // avoid unused variable warning
```
Comment (dalg24) [inline core/unit_test/TestPolicyConstruction.hpp:798]: ??

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 2]
```diff
@@ -417,8 +417,8 @@ class half_t {
     // Use non-volatile val_ref to suppress:
     // "warning: implicit dereference will not access object of type ‘volatile
     // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) *
+    auto& val_ref = const_cast<impl_type&>(val);
```
Comment (e10harvey) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:420]: is this an rvaule ref now?
```suggestion
    impl_type& val_ref = const_cast<impl_type&>(val);
```

**Review by e10harvey — COMMENTED** [review]
Thanks, @masterleinad ! Which are the faulty `half_t` operations?

File: core/unit_test/TestPolicyConstruction.hpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/TestPolicyConstruction.hpp:798]: `nvc++` still complained here using `(void)` but this syntax seems to silence the warning (on all CI checks).

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:420]: Yes, exactly, but I prefer not to write the type twice.

**Comment by masterleinad** [comment]
> Thanks, @masterleinad ! Which are the faulty `half_t` operations?

All the places where I changed `auto` to `auto &`, i.e.
- `volatile half_t& operator+=(half_t rhs) volatile`
- `volatile half_t& operator-=(half_t rhs) volatile`
- `volatile half_t& operator*=(half_t rhs) volatile`
- `volatile half_t& operator/=(half_t rhs) volatile`

for the non-CUDA case.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by e10harvey** [comment]
> All the places where I changed `auto` to `auto &`, i.e.
> 
>     * `volatile half_t& operator+=(half_t rhs) volatile`
> 
>     * `volatile half_t& operator-=(half_t rhs) volatile`
> 
>     * `volatile half_t& operator*=(half_t rhs) volatile`
> 
>     * `volatile half_t& operator/=(half_t rhs) volatile`
> 
> 
> for the non-CUDA case.

Ok, this is done in #4101.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 2 above)
Comment (PhilMiller) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:420]: I think this was changed in #4156. If not, the comments there address why this warning is real, and this fix is not legitimate

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 3]
```diff
@@ -468,8 +468,8 @@ class half_t {
     // Use non-volatile val_ref to suppress:
     // "warning: implicit dereference will not access object of type ‘volatile
     // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) /
+    auto& val_ref = const_cast<impl_type&>(val);
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:471]: This was definitely fixed in a better way in #4156

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 2 above)
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:420]: This was changed in #4156.

**Review by PhilMiller — CHANGES_REQUESTED** [review]
Please merge or rebase this PR to incorporate the changes from #4156, which I think moot all changes in `Kokkos_Cuda_Half.hpp` - the changes to it here are papering over incorrect code that the compiler legitimately warns about.

**Comment by PhilMiller** [comment]
To be clear, I'm +1 on the rest of the changes here

**Comment by masterleinad** [comment]
@PhilMiller I rebased.

**Review by PhilMiller — APPROVED** [review]

File: containers/src/Kokkos_DualView.hpp [hunk 4]
```diff
@@ -813,8 +814,17 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
+      std::enable_if_t<Dummy::impl_dualview_is_single_device::value>* = nullptr>
+  void modify() {
+    return;
+  }
+
+  template <class Dummy = DualView,
+            std::enable_if_t<!Dummy::impl_dualview_is_single_device::value>* =
+                nullptr>
   inline void modify_host() {
-    if (impl_dualview_is_single_device::value) return;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_DualView.hpp:817]: Something like
```
"/tmp/kokkos/containers/unit_tests/../src/Kokkos_DualView.hpp", line 818: warning: statement is unreachable
```

**Review by crtrott — APPROVED** [review]

**Pull request merged**